File: yetutil/src/main/java/net/yet/ui/dialogs/ListViewDialog.java
package net.yet.ui.dialogs;

import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.DialogInterface.OnClickListener;

import java.util.ArrayList;
import java.util.Collection;

/**
 * 简单字符串选择
 */
public abstract class ListViewDialog {
    private ArrayList<String> list = new ArrayList<String>(64);

    public int size() {
        return list.size();
    }

    public void addItems(String... items) {
        for (String s : items) {
            list.add(s);
        }
    }

    public void addItems(Collection<String> items) {
        for (String s : items) {
            list.add(s);
        }
    }

    public AlertDialog show(Context context) {
        return show(context, null);
    }

    public AlertDialog show(Context context, String title) {
        String[] data = list.toArray(new String[list.size()]);
        AlertDialog.Builder builder = new AlertDialog.Builder(context);
        builder.setTitle(title);
        builder.setItems(data, new OnClickListener() {

            @Override
            public void onClick(DialogInterface dialog, int which) {
                dialog.dismiss();
                onSelect(which, list.get(which));
            }
        });
        builder.setCancelable(true);
        AlertDialog dlg = builder.create();
        dlg.setCanceledOnTouchOutside(true);
        dlg.show();
        return dlg;
    }

    abstract protected void onSelect(int index, String s);
}

File: yetutil/src/main/java/net/yet/ui/dialogs/ListViewDialog.kt
package net.yet.ui.dialogs

import android.app.AlertDialog
import android.content.Context
import android.view.Gravity
import net.yet.ui.ext.dp
import java.util.*

/**
 * 简单字符串选择
 */
abstract class ListViewDialog {
	private val list = ArrayList<String>(64)
	var gravityTop = false
	var topMargin = 45


	fun size(): Int {
		return list.size
	}

	fun addItems(vararg items: String) {
		for (s in items) {
			list.add(s)
		}
	}

	fun addItems(items: Collection<String>) {
		for (s in items) {
			list.add(s)
		}
	}

	fun show(context: Context, title: String? = null): AlertDialog {
		val data = list.toTypedArray()
		val builder = AlertDialog.Builder(context)
		builder.setTitle(title)
		builder.setItems(data) { dialog, which ->
			dialog.dismiss()
			onSelect(which, list[which])
		}
		builder.setCancelable(true)
		val dlg = builder.create()
		dlg.setCanceledOnTouchOutside(true)

		onConfigDialog(dlg)
		dlg.show()
		return dlg
	}

	open fun onConfigDialog(dlg: AlertDialog) {
		if (gravityTop) {
			val lp = dlg.window.attributes
			lp.gravity = Gravity.TOP or Gravity.CENTER_HORIZONTAL
			lp.y = dp(topMargin)
		}
	}

	abstract fun onSelect(index: Int, s: String)
}
